Ugizwenayo-Divine/RICA-INC
# customer orders a product

#### What does this PR do?
- creating order 
- canceling order
- customer gets all his orders
- customer gets his one order
- admin search any order
- admin views all orders based on status
- admin views all orders of different customers
- customer pays his order either by momo or card
- verify the payed orders
- get all verified transaction
- admin delete the order
- admin is able to add a created product to best products
- admin is able to remove the product from the best products
- the admin is able to view all sold products
- the admin is able to view all available products
- the admin is able to view all best products
- the admin is able to view all normal products
#### Description of Task to be completed?
- added needed packages
- created controller
- created services
- created some validations
- made all required middlewares
#### How should this be manually tested?
- After cloning the project repo, run npm i inside the cloned repo folder
- change in cloned repo
- git fetch
- git checkout ft-customer-order-product
- npm i
- provide all required env variables (SECRET_KEY=secret API key from flutterwave)
- npm run migrations
- npm run admin to insert the admin user
- login as admin on POST `localhost:3000/api/auth/login`
- add order on POST `localhost:3000/api/orders/add`
- delete order on DELETE `localhost:3000/api/orders/:id`
- cancel the order on PATCH `localhost:3000/api/orders/cancel/:id`
- get order on GET `localhost:3000/api/orders/:id`
- get all orders by admin on GET `localhost:3000/api/orders`
- get order by customer on GET `localhost:3000/api/orders/client-orders/:id`
- get all orders by admin on GET `localhost:3000/api/orders/client-orders`
- get all orders based on status by admin on GET `localhost:3000/api/orders/status-based?status=any`
- pay the order on POST `localhost:3000/api/orders/pay/:id`
- verify the transaction on PATCH `localhost:3000/api/payment/verify/:transactionId`
- get all verified transactions by admin on GET `localhost:3000/api/payment/transactions`
- admin add a created product to best products on PATCH `localhost:3000/api/product/add-best/:id`
- admin remove the product from the best products on PATCH `localhost:3000/api/product/remove-best/:id`
- the admin view all sold products on GET `localhost:3000/api/product/sold`
- the admin view all available products on GET `localhost:3000/api/product/available`
- the admin view all best products on GET `localhost:3000/api/product/best-product`
- the admin view all normal products on GET `localhost:3000/api/product/normal-product`
#### Any background context you want to provide?
- you must have node installed locally
#### What are the relevant pivotal tracker stories?
N/A
#### Screenshots (if appropriate)
N/A
